### DISCUSSION TITLE
Better performance

### DISCUSSION THREAD

**Discussion by kaatinga** [jellydator/ttlcache#105] [Ideas]
I see that we have several issues related to performance. I would like to help with this, like to increase performance :) who can help me to go through the current solution?

**Comment by swithek** [comment]
It seems there are only two issues mentioning potential performance improvements: https://github.com/jellydator/ttlcache/issues/92 and https://github.com/jellydator/ttlcache/issues/93. I'm not sure it's possible to implement the changes proposed in #92 because `Get()` depends on normal mutex locking when extending the TTL of an item. Of course, this feature is optional and can be disabled, but that would mean that we would need to determine which type of lock to use based on that option and I'm not sure whether that's the best way to address this (though I might think otherwise when I see the implementation).

**Reply by kaatinga** [reply]
Is there nothing besides the https://github.com/jellydator/ttlcache/issues/92 and https://github.com/jellydator/ttlcache/issues/93 on mind? I can look at https://github.com/jellydator/ttlcache/issues/93

**Reply by swithek** [reply]
It might also be a good idea to expand the list of [benchmark tests](https://github.com/jellydator/ttlcache/blob/master/bench/bench_test.go) so that we could see how these performance improvements affect everything.

**Reply by gozeloglu** [reply]
Yes, you are right. `Get` depends on ttl update. Maybe the `MGetWithoutTouchingOnHit` (or something similar name) method can be implemented. It will be worked when `disableTouchOnHit = true`. If it is `false`, the `MGetWithoutTouchingOnHit` returns nil or error.

**Reply by KyleSanderson** [reply]
> It seems there are only two issues mentioning potential performance improvements: #92 and #93. I'm not sure it's possible to implement the changes proposed in #92 because `Get()` depends on normal mutex locking when extending the TTL of an item. Of course, this feature is optional and can be disabled, but that would mean that we would need to determine which type of lock to use based on that option and I'm not sure whether that's the best way to address this (though I might think otherwise when I see the implementation).

This is an incredible overhead that I don't think has been fully exposed to you. regex compilation used to dominate my profiles, now it's the expiration queue which is ironic, hilarious, and "what, really" all at the same time.

Are you not seeing the same thing? There has to be a better way to handle this (a ticker and just updating the items under a mutex) would be better in this context. Is there a better way to handle this?

![{58E1DE66-0CBB-4882-BFD2-F64A5D0FBE8B}](https://github.com/user-attachments/assets/ec253f11-0f22-4269-93a1-2a1f847d4c4c)

**Reply by swithek** [reply]
It would be nice if you could share your configuration options, data volume, and use case @KyleSanderson. It's hard to understand the full picture without that.

> This is an incredible overhead that I don't think has been fully exposed to you.

That's why issues and discussions exist, we improve the package based on that.

> Is there a better way to handle this?

Always. Suggestions and PRs are more than welcome, but please refrain from snarky comments, they help no one.

**Comment by ReneKroon** [comment]
An additional change could be an atomic  per item CAS (compare and swap)
data structure as backing store. Better than sharding, which is an O(1)
solution.

Op ma 5 jun. 2023 08:06 schreef Gökhan Özeloğlu ***@***.***>:
